From 5c345bb51000c36f92d4a249801228d030cf81b0 Mon Sep 17 00:00:00 2001
From: Linus Torvalds <torvalds@linux-foundation.org>
Date: Fri, 5 Jan 2018 16:26:00 -0800
Subject: [PATCH 27/60] loop: fix concurrent lo_open/lo_release
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

范龙飞 reports that KASAN can report a use-after-free in __lock_acquire.
The reason is due to insufficient serialization in lo_release(), which
will continue to use the loop device even after it has decremented the
lo_refcnt to zero.

In the meantime, another process can come in, open the loop device
again as it is being shut down. Confusion ensues.

Reported-by: 范龙飞 <long7573@126.com>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Git-commit: ae6650163c66a7eff1acd6eb8b0f752dcfa8eba5
Git-repo: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git
Signed-off-by: Srinivasa Rao Kuppala <srkupp@codeaurora.org>

Conflicts:
	drivers/block/loop.c

Signed-off-by: Dheeraj Kumar Chaudhary <dheech@codeaurora.org>
Change-Id: I965ddcdc2a9d27fd56a54d2886b7c3861c21b05f
CVE-2018-5344
Signed-off-by: Kevin F. Haggerty <haggertk@lineageos.org>
---
 drivers/block/loop.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/block/loop.c b/drivers/block/loop.c
index 462fd182180..d84f42190af 100644
--- a/drivers/block/loop.c
+++ b/drivers/block/loop.c
@@ -1505,9 +1505,8 @@ out:
 	return err;
 }
 
-static int lo_release(struct gendisk *disk, fmode_t mode)
+static int __lo_release(struct loop_device *lo)
 {
-	struct loop_device *lo = disk->private_data;
 	int err;
 
 	mutex_lock(&lo->lo_ctl_mutex);
@@ -1537,6 +1536,15 @@ out_unlocked:
 	return 0;
 }
 
+static int lo_release(struct gendisk *disk, fmode_t mode)
+{
+	int err;
+	mutex_lock(&loop_index_mutex);
+	err =__lo_release(disk->private_data);
+	mutex_unlock(&loop_index_mutex);
+	return err;
+}
+
 static const struct block_device_operations lo_fops = {
 	.owner =	THIS_MODULE,
 	.open =		lo_open,
-- 
2.17.1

